aspnet/fundamentals/localization.rst

```diff
@@ -1,8 +1,302 @@
+  
+Localization middleware
+-----------------------------------
+
+The current culture on a request is set in the localization :doc:`/fundamentals/middleware`. The localization middleware is enabled in the ``Configure`` method of *Startup.cs* file.
+
+.. literalinclude:: localization/sample/StartupConfig.cs
+  :language: c#
+  :linenos:
+  :lines: 106-107, 136-159
+  :dedent: 6
+
+
+``UseRequestLocalization`` initializes a ``RequestLocalizationMiddleware`` object. The ``Invoke`` method of the `RequestLocalizationMiddleware <https://github.com/aspnet/Localization/blob/dev/src/Microsoft.AspNetCore.Localization/RequestLocalizationMiddleware.cs>`__ class is called by the ASP.NET Core pipeline. The ``Invoke`` method enumerates the ``RequestCultureProviders`` list, and returns the first non-null provider. The default providers come from the ``RequestLocalizationOptions`` class:
+
+#. ``QueryStringRequestCultureProvider``
+#. ``CookieRequestCultureProvider``
+#. ``AcceptLanguageHeaderRequestCultureProvider``
+
+The default list goes from most specific to least specific. Later in the article I'll show how you can change the order and even add a custom localization provider. If there are no non-null providers, the ``DefaultRequestCulture`` is used.
+    
+QueryStringRequestCultureProvider
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Adding a query string to the URL is useful for debugging and testing code. Unless you change the ``RequestCultureProviders`` list, adding a query string to the URL will always win as the localization provider. You pass the query string parameters ``culture`` and ``ui-culture``. The following example sets the specific culture (language and region) to Spanish/Mexico:
+
+``http://localhost:5000/?culture=es-MX&ui-culture=es-MX``
+
+If you only pass in one of the two (``culture`` or ``ui-culture``), the query string provider will set both values using the one you passed in. For example, setting just the culture will set both the ``Culture`` and the ``UICulture``:
+
+``http://localhost:5000/?culture=es-MX``
+
+CookieRequestCultureProvider
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Production app will often provide a mechanism to set the culture with the ``ASPNET_CULTURE`` cookie. The sample app provides code to set the cookie for French and English cultures.
+
+The Accept-Language HTTP header
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The Accept-Language HTTP header from a browser request is not an infallible way to detect the user's preferred language. This settings indicates what the browser has been set to send or has inherited from the underlying operating system. In a production app you should include a way for users to set the culture with the ``ASPNET_CULTURE`` cookie.
+
+Use ``RequestLocalizationOptions`` to add or remove localization providers. You can use the ``Remove`` or ``RemoveAt`` method to remove a provider. For example, the following code removes the ``QueryStringRequestCultureProvider``:
```

Comment:
It removes the *first* provider, which just *happens* to be at this point in time `QueryStringRequestCultureProvider`, so I recommend changing this to just say "removes the first." On the other hand, this could probably be omitted entirely because people ought to know how to clear collections.